Only use the "polynomial" definition of ff() and rf() if the input is a Poly

asmeurer opened the pull request:
Fixes #11105

See also #8941.

CC @postvakje

asmeurer committed:
Only use the "polynomial" definition of ff() and rf() if the input is a Poly

Fixes #11105

See also #8941.

jksuom commented on sympy/functions/combinatorial/factorials.py:
```diff
@@ -480,37 +481,38 @@ def eval(cls, x, k):
-                            return reduce(lambda r, i: r*(x + i), range(0, int(k)), 1)
+                        if isinstance(x, Poly):
+                            gens = x.gens
+                            if len(gens) > 1 or x.degree() <= 1:
```
I'm not sure that these cases need to be singled out. I think that this should be called only with univariate polynomials, multivariate being considered as errors. The linear case (`a*y + b, a != 0`) could be handled together with the nonlinear polynomials. Constant polynomials may not give what had been expected (a power instead of a factorial), but those are not strictly speaking univariate either.

asmeurer replied:
I was just going with what was already here. I'm not really sure if this is how it should work.

asmeurer replied:
So you are suggesting for `len(gens) > 1` to raise ValueError?

jksuom replied:
I think that could be done. If there are several symbols, then the others could belong to the coefficient domain with only one primary symbol left as generator.

jksuom commented on sympy/functions/combinatorial/factorials.py:
```diff
@@ -480,37 +481,38 @@ def eval(cls, x, k):
+                            if len(gens) > 1 or x.degree() <= 1:
+                                return reduce(lambda r, i: r*(x + i), range(0, int(k)), 1)
+                            else:
+                                v = gens[0]
```
If len(gens) == 1 is checked, then this would also exist.

asmeurer replied:
You can't have a Poly with 0 gens, as far as I know. Or am I misunderstanding what you are saying here?

jksuom replied:
There could exist PolyElements with no generators though those are not likely to appear in this context. I was only thinking that if len(gens) == 1 is explicitly checked (instead of len(gens) > 1), then that (remote) possibility would also be taken care of.

asmeurer commented:
I also see that Poly.subs does something stupid.

```
>>> Poly(x, x).subs(x, x + 1)
Poly((x + 1), x + 1, domain='ZZ')
```

So I've changed it to use `shift` instead.

asmeurer committed:
Change rf and ff to raise ValueError for Poly with more than one generator

Also use shift() instead of subs(), as Poly(..).subs(x, x + 1) adds a new
generator x + 1.

asmeurer commented:
Going to merge so I can start the release.
